scipy/stats/_qmc.py

```diff
@@ -883,6 +977,188 @@ def reset(self):
         return self
 
 
+class OptimalDesign(QMCEngine):
+    """Optimal design.
+
+    Optimize the design by doing random permutations to lower the centered
+    discrepancy [1]_.
+
+    The specified optimization `method` is used to select a new set of
+    permutations to perform. If `method` is None, *basinhopping* optimization
+    is used. `niter` set of permutations are performed.
+
+    Centered discrepancy-based design shows better space filling robustness
+    toward 2D and 3D subprojections. Distance-based design shows better space
+    filling but less robustness to subprojections [2]_.
+
+    Parameters
+    ----------
+    d : int
+        Dimension of the parameter space.
+    start_design : array_like (n, d), optional
+        Initial design of experiment to optimize. `OrthogonalLatinHypercube`
+        is used to generate a first design otherwise.
+    niter : int, optional
+        Number of iterations to perform. Default is 1.
+    method : callable ``f(func, x0, bounds)``, optional
+        Optimization function used to search new samples. Default to
```

Comment:
Currently, `basinhopping` does not appear to be performing really well; it seems to explore a very small portion of the possible swaps before terminating. The output design is very similar to the input after one iteration, and they differ only in small-number rows/cols. (e.g. use `np.where(before != after)`to see the differences).

Whatever optimizer you do end up with, I'd like to see it tested head-to-head against random permutations to see how the discrepancy goes down as a function of time for each. I'm not sure if the overhead of an optimizer is really going to be worth it, since the objective function keeps changing on them.

Reply:
This is great. Thanks for working hard on this! I will definitely use your script to test `dual_annealing`. I will also vary the dimension. Not sure you remember this, but the original code I had for the PR on QMC had 2 strategies. It was first checking the complexity of the problem (just calculating the total number of possible combinations). And if the complexity was bellow a threshold, it was going through all the possible permutations. Otherwise it would use basinhopping.
I am curious to see how `dual_annealing` performs here. Everyone uses it in the literature so I hope to see better results 🤷‍♂️

Reply:
2-opt is a little different than all possible permutations of the LHS. It is a greedy version of that idea. In the process of making all possible exchanges between two rows for the same column, you don't generate all possible Latin Hypercubes. But yes, best of e.g. 1000 randomly generated Latin Hypercubes and best of all possible Latin Hypercubes (if the problem is small) are also good ideas.

Reply:
In the first version I also had a mode with a simple loop to generate n LHS and take the best. At the time we discarded this as we said it was too simple and anyone could do this without problem. This could be part of an example to show how `discrepancy` can be useful.

Reply:
I went again through the main article: `Jin et al., "An efficient algorithm for constructing optimal design
       of computer experiments."
       Journal of Statistical Planning and Inference, 2005.`. Actually I should have something along what you proposed here.

There are 2 loops. The outer loop is the annealing part and the inner loop consist of doing random permutations and updating the best sample (similar to what you do). The two loops are linked through the temperature, it is used to select or discard a new sample. I think the current implementation of `dual_annealing` does not give access to the `temperature`. Maybe I could add a flag in order to allow passing it to the function (I think here https://github.com/scipy/scipy/blob/374e5bfec63e530c34514a08d9311352138182f3/scipy/optimize/_dual_annealing.py#L297).

cc @andyfaff

I updated the algorithm to make it a bit closer to the paper (without the temperature link for now). And it performs better (using your tests), especially for large dimensions.

In the end, if I manage to replicate the algorithm with the temperature and such, I would remove the `method` as it would not really make sense.

Reply:
I implemented the algorithm from the article. It seems good. I still need to check the performance agains the naive implementation and tests your `two_opt`. But the before and after is quite different now. For a design of size (10, 400), `np.where(before != after)` gives 1548 changes.

Reply:
@mdhaber for reference, I updated the description of the PR with the flowchart of the algorithm.

Reply:
What are the results of your tests of its effectiveness in a given amount of time (or time rate of discrepancy reduction)?

Reply:
If we reuse the best design and iterate like this, we obtain roughly the same performance with the new ESE. 

```python
original: 0.001231782193903852
basinhopping: 0.0011975526751670795
basinhopping: 0.0011950900444532842
...
basinhopping: 0.0011950900444532842
basinhopping: 0.0011950900444532842
basinhopping: 0.0011950900444532842
basinhopping: 0.0011950900444532842
swapper: 0.0003416202432611204
swapper: 0.0002696987661519046
...
swapper: 0.0001945242170817707
swapper: 0.00019050135380038746
ESE: 0.00025261993981051667
ESE: 0.00020622234542688744
```

But if we just do one shot with the same number of iterations, I have better results with ESE.
```python
original: 0.001231782193903852
basinhopping: 0.0011476753783927762
swapper: 0.0002172170249121752
ESE: 0.00018944240846363591
```

Reply:
For sure the previous strategy was just not working. Now the new one seem to be doing something better. Still maybe I have a bug in my implementation of the paper (like I was not entirely sure about the sample to give to the inner loop).

FYI, to get the results above I just took a previous commit (a72903b873d2831decafb3955f4205db13e5b0af) and copied the old `OptimalDesign` so that I had the 2 classes available.

Reply:
Are there are results in the paper you can try to reproduce that might hint at whether it's implemented correctly?
Also, it would be good to have stronger evidence that the method worth the added complexity. For instance, plots of discrepancy as a function of _time_ (as iterations could take different amounts of time) for a few different sizes/dimensions would make the decision easier.

Reply:
I will have a look at both, thanks Matt 😃

Reply:
I had a look at some numerical results from `Damblin et al., "Numerical studies of space filling designs: optimization of Latin Hypercube Samples and subprojection properties." Journal of Simulation, 2013.`. I could only check the convergence for C2 (canonical-subspaces: I only selected one subspace here) and minimum spanning tree (I had this code already). For the other figures (same with the paper from Jin2015) I would need to code another criterion: `phiP` and allow to change the metric used during the optimization.

The results seem to follow the findings from figures 9 and 10 of the paper.

Side note on the metrics: I am thinking of doing a separate PR to add MST and other distance variantes. What do you think?

<details>
  <summary>Reproducible code.</summary>

```python
import numpy as np
from scipy.stats import qmc
from scipy.spatial.distance import cdist
from scipy.sparse.csgraph import minimum_spanning_tree
import matplotlib.pyplot as plt


def l1_norm(sample):
    l1 = cdist(sample, sample, 'cityblock')
    return np.min(l1[l1.nonzero()])


# mindist
def l2_norm(sample):
    l2 = cdist(sample, sample)
    return np.min(l2[l2.nonzero()])


def linf_norm(sample):
    linf = cdist(sample, sample, 'chebyshev')
    return np.min(linf[linf.nonzero()])


def mst(sample, plot=False):
    """Minimum Spanning Tree criterion.

    MST is used here as a discrepancy criterion.
    Comparing two different designs: the higher the mean,
    the better the design is in terms of space filling.

    Parameters
    ----------
    sample : array_like (n, d)
        The sample to compute the discrepancy from.
    Returns
    -------
    mean, var : float
        Mean and standard deviation of the MST.
    """
    sample = np.asarray(sample)

    dist = cdist(sample, sample)
    mst = minimum_spanning_tree(dist)

    if (sample.shape[1] == 2) and plot:

        edges = np.where(mst.toarray() > 0)
        edges = np.array(edges).T

        fig, ax = plt.subplots()
        for edge in edges:
            ax.plot(sample[edge, 0], sample[edge, 1], c='k')
        ax.scatter(sample[:, 0], sample[:, 1])

        plt.show()

    return mst.data.mean(), mst.data.std()


qmc_gen = qmc.LatinHypercube(2)
doe = qmc_gen.random(20)
mst(doe, plot=True)


# figure 10: MST mean  0.7->0.8 ; std. 0.082 -> 0.065
n, d = 100, 10
qmc_gen = qmc.LatinHypercube(d)
doe = qmc_gen.random(n)
print(mst(doe))

for i in range(10):
    lhs_gen = qmc.OptimalLatinHypercube(d, doe, niter=1)
    doe = lhs_gen.random(n)
    print(mst(doe))


# figure 11: c2 disc 0.017 non optim and <0.01 optim
# C2 here need ** 0.5, it's a known difference due to the
# interpretation of the term C2 which needs or not to be squared.
# Our implementation is consistent with numerical results from the base literature.
n, d = 100, 10
qmc_gen = qmc.LatinHypercube(d)
doe = qmc_gen.random(n)
print(qmc.discrepancy(doe[:, :2]) ** 0.5)

for i in range(5):
    lhs_gen = qmc.OptimalLatinHypercube(d, doe, niter=1)
    doe = lhs_gen.random(n)
    # print(l2_norm(doe), l1_norm(doe), linf_norm(doe), qmc.discrepancy(doe[:, :2]) ** 0.5)
    print(qmc.discrepancy(doe[:, :2]) ** 0.5)
```
</details>

Reply:
> Side note on the metrics: I am thinking of doing a separate PR to add MST and other distance variantes. What do you think?

Interesting metric! I don't have any experience with it, but sounds useful.

As for the results, yes, there is some evidence that the match the paper. (Makes me wonder how simple random permutations would fare, but I suppose I can check that on my own when I am looking at this again.)

Did you happen to look into discrepancy as a function of time, or still working on that?

Reply:
> Interesting metric! I don't have any experience with it, but sounds useful.

I will put this on my list then, thanks.

> As for the results, yes, there is some evidence that the match the paper. (Makes me wonder how simple random permutations would fare, but I suppose I can check that on my own when I am looking at this again.)
> 
> Did you happen to look into discrepancy as a function of time, or still working on that?

Not yet, just need to find a bit of time for it 😅   But I also would like to be certain that I did not miss anything here.

Reply:
Ok here is the plot with respect to time, and yes the not fancy method works better... I am sending an email to the authors of these papers to see if I have done something wrong. Otherwise I will just go with your idea. Later on we can add a `method` param if somehow someone finds it 😅

![olhs_dummy](https://user-images.githubusercontent.com/23188539/117413563-685d0500-af16-11eb-874a-1dfea1e03976.png)

Reply:
@jschueller if you have some time maybe you can help here 😃  Maybe I have a very stupid bug here like mixing up some variables??

(Just careful about licensing, don't speak about how it's coded in OT)

Reply:
I just got one reply from G. Damblin and it seems like the algorithm is very difficult to parametrize. He is saying that it should be better in high dimension, but was not that convincing to me. So without further help and dedicated time to research this, I think that this approach will not work for us...

What I propose now is to use your `swapper` @mdhaber and if someone manages to code ESE and show a real gain (and I am now skeptical about this after having tried... thanks for all your questions which lead me to test this more by the way 😃 ) we can revisit.

(I would not try to push this before the release so we don't need to stress).